File: src/main/java/dev/anhcraft/enc/filters/FilterAssistant.java
package dev.anhcraft.enc.filters;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import dev.anhcraft.enc.filters.entity.NPCFilter;
import dev.anhcraft.enc.filters.entity.ShopkeeperFilter;
import org.bukkit.entity.Entity;

public class FilterAssistant {
    private static final Multimap<Class<?>, ObjectFilter<?>> FILTERS = new HashMultimap<>();

    static {
        FILTERS.put(Entity.class, new NPCFilter());
        FILTERS.put(Entity.class, new ShopkeeperFilter());
    }

    @SuppressWarnings("unchecked")
    public static <T> boolean anyMatch(T object, Class<?> clazz){
        var f = FILTERS.get(clazz);
        for(var of : f){
            var of_ = (ObjectFilter<T>) of;
            if(of_.check(object)) return true;
        }
        return false;
    }
}

File: src/main/kotlin/dev.anhcraft.enc/utils/FilterAssistant.kt
package dev.anhcraft.enc.utils

import com.google.common.collect.ArrayListMultimap
import org.bukkit.entity.Entity
import org.bukkit.entity.LivingEntity
import java.util.function.Predicate

object FilterAssistant {
    private val FILTERS = ArrayListMultimap.create<Class<*>, Predicate<*>>()

    init {
        FILTERS.put(Entity::class.java, Predicate<Entity> { o -> o.hasMetadata("NPC") })
        FILTERS.put(Entity::class.java, Predicate<Entity> { o -> o.hasMetadata("shopkeeper") })
        FILTERS.put(Entity::class.java, Predicate<Entity> { o -> o !is LivingEntity })
    }

    @JvmStatic
    fun <T> anyMatch(o: T, clazz: Class<T>): Boolean {
        val f = FILTERS.get(clazz)
        for (of in f) {
            if ((of as Predicate<T>).test(o)) return true
        }
        return false
    }
}
